components/dropdown-menu/index.js

```diff
@@ -60,47 +79,58 @@ class DropdownMenu extends wp.element.Component {
 	}
 
 	findActiveIndex() {
-		if ( this.menuRef ) {
-			const menu = findDOMNode( this.menuRef );
+		if ( this.menuNode ) {
 			const menuItem = document.activeElement;
-			if ( menuItem.parentNode === menu ) {
-				return findIndex( menu.children, ( child ) => child === menuItem );
+			if ( menuItem.parentNode === this.menuNode ) {
+				return findIndex( this.menuNode.children, ( child ) => child === menuItem );
 			}
 			return -1;
 		}
 	}
 
 	focusIndex( index ) {
-		if ( this.menuRef ) {
-			const menu = findDOMNode( this.menuRef );
+		if ( this.menuNode ) {
 			if ( index < 0 ) {
-				menu.previousElementSibling.focus();
+				this.menuNode.previousElementSibling.focus();
 			} else {
-				menu.children[ index ].focus();
+				this.menuNode.children[ index ].focus();
 			}
 		}
 	}
 
 	focusPrevious() {
 		const i = this.findActiveIndex();
-		const prevI = i <= -1 ? -1 : i - 1;
+		const maxI = this.props.controls.length - 1;
+		const prevI = i <= 0 ? maxI : i - 1;
 		this.focusIndex( prevI );
 	}
 
 	focusNext() {
 		const i = this.findActiveIndex();
 		const maxI = this.props.controls.length - 1;
-		const nextI = i >= maxI ? maxI : i + 1;
+		const nextI = i >= maxI ? 0 : i + 1;
 		this.focusIndex( nextI );
 	}
 
+	handleKeyUp( event ) {
+		// TODO: find a better way to isolate events on nested components see GH issue #1973.
+		if ( this.state.open ) {
+			/*
+			 * VisualEditorBlock uses keyup to deselect the block. When the menu is
+			 * open we need to stop propagation of keyup after Escape has been pressed
+			 * to close the menu, otherwise the whole block toolbar will disappear.
+			 * When the menu is closed, Escape will make the toolbar disappear as intended.
+			 */
+			event.stopPropagation();
+		}
+	}
+
 	handleKeyDown( keydown ) {
 		if ( this.state.open ) {
 			switch ( keydown.keyCode ) {
 				case ESCAPE:
 					keydown.preventDefault();
 					keydown.stopPropagation();
-					this.closeMenu();
 					const node = findDOMNode( this );
 					const toggle = node.querySelector( '.components-dropdown-menu__toggle' );
 					toggle.focus();
```

Comment:
I'm finding that occasionally pressing escape when navigating buttons will clear focus altogether (`document.activeElement === document.body`). Not sure if that's specific to your changes here though.

macOS Chrome 59.0.3071.115